src/components/hub/elements/MessagesList.js

```diff
@@ -56,74 +55,42 @@ export default function MessagesList({ scrollRef }) {
-        cid: selectedCID,
-        limit: +process.env.REACT_APP_MESSAGES_COUNT_TO_PRELOAD,
-        updated_at: { lt: messages[0].created_at },
-      })
-      .then((arr) => {
-        if (!arr.length) {
-          needToGetMoreMessage.current = false;
-          return;
-        }
-
-        const messagesIds = arr.map((el) => el._id).reverse();
-        needToGetMoreMessage.current = !(
-          messagesIds.length < +process.env.REACT_APP_MESSAGES_COUNT_TO_PRELOAD
-        );
-
-        updateParticipantsFromMessages(arr);
-        dispatch(addMessages(arr));
-        dispatch(
-          upsertChat({
-            _id: selectedCID,
-            messagesIds: [
-              ...new Set([...messagesIds, ...messages.map((el) => el._id)]),
-            ],
-            activated: true,
-          })
-        );
-        const mAttachments = {};
-        for (let i = 0; i < arr.length; i++) {
-          const attachments = arr[i].attachments;
-          if (!attachments) {
-            continue;
-          }
-          attachments.forEach((obj) => {
-            const mAttachmentsObject = mAttachments[obj.file_id];
-            if (!mAttachmentsObject) {
-              mAttachments[obj.file_id] = {
-                _id: arr[i]._id,
-                ...obj,
-              };
-              return;
-            }
-
-            const mids = mAttachmentsObject._id;
-            mAttachments[obj.file_id]._id = Array.isArray(mids)
-              ? [arr[i]._id, ...mids]
-              : [arr[i]._id, mids];
-          });
-        }
-
-        if (Object.keys(mAttachments).length > 0) {
-          DownloadManager.getDownloadFileLinks(mAttachments).then((msgs) =>
-            dispatch(upsertMessages(msgs))
-          );
-        }
-
-        if (Object.keys(mAttachments).length > 0) {
-          DownloadManager.getDownloadFileLinks(mAttachments).then((msgs) => {
-            const messagesToUpdate = msgs.flatMap((msg) => {
-              const mids = Array.isArray(msg._id) ? msg._id : [msg._id];
-              return mids.map((mid) => ({ ...msg, _id: mid }));
-            });
+    const processMessages = (arr) => {
+      if (!arr.length) {
+        needToGetMoreMessage.current = false;
+        return;
+      }
 
-            dispatch(upsertMessages(messagesToUpdate));
-          });
-        }
-      });
+      const messagesIds = arr.map((el) => el._id).reverse();
+      needToGetMoreMessage.current = !(
+        messagesIds.length < +process.env.REACT_APP_MESSAGES_COUNT_TO_PRELOAD
+      );
+
+      updateParticipantsFromMessages(arr);
+      dispatch(addMessages(arr));
+      dispatch(
+        upsertChat({
+          _id: selectedCID,
+          messagesIds: [
+            ...new Set([...messagesIds, ...messages.map((el) => el._id)]),
+          ],
+          activated: true,
+        })
+      );
+
+      const mAttachments = messagesService.processAttachments(arr);
+      messagesService.retrieveAttachmentsLinks(mAttachments);
```

Comment:
I have the very last comment

This code is duplicated in `messageService.syncData`, hence we can refactor it a bit to reduce code duplication.

One possible way to do so is to move this code directly into  `messageService.retrieveMessages`, e.g.


```
  async retrieveMessages(params) {
    const messagesDB = await indexedDB.getMessages(params);

    if (messagesDB.length >= params.limit) return messagesDB;
    const messagesAPI = await api.messageList(params);
    await indexedDB.insertManyMessages(messagesAPI);
    
   const messagesIds = messages.map((el) => el._id).reverse();

   const jointMessages =  [...messagesDB, ...messagesAPI]; // do we also need to pass here ids from redux?
   
   // save to redux store 
   store.dispatch(addMessages(jointMessages));
   store.dispatch(
      upsertChat({ _id: this.currentChatId, messagesIds, activated: true })
   );

   const mAttachments = messagesService.processAttachments(arr);
   messagesService.retrieveAttachmentsLinks(mAttachments);
  }
```

WDYT?